lib/veri.ml

```diff
@@ -194,66 +190,57 @@ class ['a] t arch dis =
-        returned and no load event is emitted.
-        Otherwise searching continues as written above for [resolve_addr],
-        with emitting memory load event. *)
-    method! eval_load ~mem ~addr endian size =
-      SM.get () >>= fun ctxt ->
-      self#eval_exp addr >>= fun addr_res ->
-      match value addr_res with
-      | Bil.Bot | Bil.Mem _ -> SM.return addr_res
-      | Bil.Imm addr' ->
-        match find_mem_store (self_events ctxt) (same_addr addr') with
-        | Some mv -> self#eval_exp (Bil.int (Move.data mv))
-        | None ->
-          self#resolve_addr mem addr endian size >>= fun r ->
-          match value addr_res, value r with
-          | Bil.Imm addr, Bil.Imm data ->
-            self#update_event (create_mem_load addr data) >>= fun () ->
+  method! eval_store ~mem ~addr data endian size =
+    super#eval_store ~mem ~addr data endian size >>= fun r ->
+    self#eval_exp addr >>= fun addr ->
+    self#eval_exp data >>= fun data ->
+    match value addr, value data with
+    | Bil.Imm addr, Bil.Imm data ->
+      SM.update (fun c -> c#discard_event (is_previous_store addr)) >>= fun () ->
+      let ev = create_mem_store addr data in
+      self#update_event ev >>= fun () -> SM.return r
+    | _ -> SM.return r
+
+  (** [find_addr_data addr] - returns a move data, bound with [addr].
+      Sequence of searches is the following:
+      1) among store events that occured at current step, in the same context,
+         with the same address; we don't need to emit event in this case
+      2) among load events that occured at current step, in the same context,
+         with the same address; will emit memory_load event
+      3) among load events that occures at current step, in other context,
+         with the same address; will emit memory_load event *)
+  method private find_addr_data addr =
+    SM.get () >>= fun ctxt ->
+    let searches = [
+      (fun () -> find_mem_store (self_events ctxt)  (same_addr addr)), false;
+      (fun () -> find_mem_load  (self_events ctxt)  (same_addr addr)), true;
+      (fun () -> find_mem_load  (other_events ctxt) (same_addr addr)), true;
+    ] in
+    SM.return (find_move_data searches)
+
+  method! eval_load ~mem ~addr endian size =
+    SM.get () >>= fun ctxt ->
+    self#eval_exp addr >>= fun addr_res ->
+    match value addr_res with
+    | Bil.Bot | Bil.Mem _ -> super#eval_load ~mem ~addr endian size
+    | Bil.Imm addr' ->
+      self#find_addr_data addr' >>= function
+      | None ->
+        begin
+          super#eval_load ~mem ~addr endian size >>= fun r ->
+          match value r with
+          | Bil.Imm data ->
+            self#update_event (create_mem_load addr' data) >>= fun () ->
             SM.return r
-          | _ ->  SM.return r
+          | _ -> SM.return r
+        end
+      | Some (data, need_emit) ->
+        super#eval_store ~mem ~addr (Bil.int data) endian size >>= fun _ ->
```

Comment:
this may lead to a non-normalized BIL, it is safe to do `super#eval_exp @@ Bil.load ~mem ~addr (Bil.int data) endian size`